test/decl/func/dynamic_self.swift

```diff
@@ -412,4 +412,88 @@ class SelfOperator {
 func useSelfOperator() {
   let s = SelfOperator()
   _ = s + s
-}
\ No newline at end of file
+}
+
+// These references to Self are now possible
+
+class A<T> {
+  let b: Int
+  required init(a: Int) {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    b = a
+  }
+  static func z() {
+    print("\(Self.self).\(#function)")
+  }
+  class func y() {
+    print("\(Self.self).\(#function)")
+    Self.z()
+  }
+  func x() -> A? {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    Self.z()
+    return Self.init(a: 77)
+  }
+}
+
+class B: A<Int> {
+  let a: Int
+  required convenience init(a: Int) {
+    print("\(Self.self).\(#function)")
+    self.init()
+  }
+  init() {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    Self.z()
+    a = 99
+    super.init(a: 88)
+  }
+  override class func y() {
+    print("override \(Self.self).\(#function)")
+  }
+}
+
+struct S2 {
+  let x = 99
+  static func x() {
+    Self.y()
+  }
+  static func y() {
+    print("y()")
+  }
+  func foo(a: [Self]) -> Self? {
+    Self.x()
+    return Self.init() as Self
+  }
+}
+
+struct S3<T> {
+  let x = 99
+  static func x() {
+    Self.y()
+  }
+  static func y() {
+    print("y()")
+  }
+  func foo(a: [Self]) -> Self? {
```

Comment:
You should add some tests with Self in non-return position of a method signature of a class as well. You'll then need to add code to ban that case, hopefully re-using the existing logic for banning `Self` in non-covariant position in protocol requirements when used as an existential.

Reply:
Not sure I follow you here. There is a test case that tests for Self in non return position of a class and it seems to work.

Reply:
Try this:

```
class C {
  func foo(_ s: Self) -> Self {
    return s
  }
}
```

That won't work because the type of s is C and not `DynamicSelfType<C>`. However if you parse `Self` in type position as DynamicSelfType for classes, you have to be careful not to allow it in invalid positions. Basically for a class, you can have Self inside a method body, or in the return type of a method, but not in the parameters of a method.

The reason is that method overrides must be covariant in Swift. For example,

```
class Base {
  func foo(_ s: Self) -> Self {
    // here I expect s and self have the same dynamic type
    return s
  }
}

class Derived : Base {}
```

We expect to be able to write:

```
let b: Base = Base.foo(Base())
let d: Derived = Derived.foo(Derived())
```

But then what happens if you do:

```
let b: Base = Derived()
let bb = b.foo(Base())
```

This in fact would introduce a type safety hole. The solution is to then not allow Self in a method signature where it is not already allowed today in Swift (however your change still makes sense for allowing Self to appear as a type or an expression in the method body, and Self anywhere in the type of struct and enum methods is OK!)

Reply:
see below